src/providers-registry/providers-registry-contract.ts

```diff
@@ -0,0 +1,270 @@
+    if (manifestData === undefined) {
+      throw new ContractError("Manifest data not set.");
+    }
+
+    if (manifestData.manifest === undefined) {
+      throw new ContractError("Manifest not set.");
+    }
+
+    if (manifestData.changeMessage === undefined || manifestData.changeMessage.length === 0) {
+      throw new ContractError("Change message is not set.");
+    }
+
+    // TODO: manifest properties validation
+
+    // note: this is safe, as manifests array is initialized when provider is added
+    providers[manifestProviderId].manifests.push(
+      {
+        uploadBlockHeight: blockHeight,
+        manifest: manifestData.manifest,
+        changeMessage: manifestData.changeMessage,
+        lockedHours: addManifestData.lockedHours || 0
+      }
+    );
+
+    trace("STATE", state);
+
+    return {state};
+  }
+
+  if (input.function === "add-provider-admin") {
+    const data = input.data as AddProviderAdminData;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+    checkPrivileges(caller, data.providerId);
+
+    // note: this is safe, as adminsPool array is initialized when provider is added
+    providers[data.providerId].adminsPool.push(...data.admins);
+
+    trace("STATE", state);
+
+    return {state};
+  }
+
+  // admin-only: switch trace state
+  if (input.function === "switch-trace") {
+    checkIsContractAdmin(caller);
+    state.trace = !state.trace;
+
+    trace("END STATE", state);
+
+    return {state}
+  }
+
+  if (input.function === "add-contract-admins") {
+    const data = input.data as AddContractAdmins;
+    if (data.admins === undefined || data.admins.length === 0) {
+      throw new ContractError("New contract admins not defined");
+    }
+    checkIsContractAdmin(caller);
+    state.contractAdmins.push(...data.admins);
+
+    return {state }
+  }
+
+
+  /* STATE READ_ONLY ACTIONS */
+  // admin-only: returns last uploaded providers-registry - ie. last element in the manifests array
+  if (input.function === "provider-data") {
+    const data = input.data as GetProviderData;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+
+    return {
+      result: {
+        provider: providers[data.providerId]
+      }
+    }
+  }
+
+  if (input.function === "active-manifest") {
+    const data = input.data as GetProviderManifest;
+    checkProviderId(data.providerId);
+    checkProviderExists(data.providerId);
+
+    const manifests = getManifestsForCaller(caller);
+    trace("Block height: ", blockHeight);
+    const manifest = manifests.slice().reverse().find((manifest) => {
+      trace("Checking manifest: ", manifest);
+      trace("Checking lockedBlocks: ", manifest.lockedHours * BLOCKS_IN_HOUR);
+      return manifest.uploadBlockHeight + (manifest.lockedHours * BLOCKS_IN_HOUR) <= blockHeight;
+    }) || manifests[0];
+
+    trace("RESULT", manifest);
+
+    return {result: {manifest}}
+  }
+
+  function checkIsContractAdmin(idWallet: string) {
```

Comment:
It's a naming condition, what do you think about AssertIsContractAdmin or MustBeContractAdmin

Reply:
Well, I'm used to this naming convention:
1. isContractAdmin - returns boolean, but does not throw exception
2. checkContractAdmin - checks condition and throws an exception if not true (sth. like "checkNotNull" or "checkArgument" from Guava's Preconditions API).

Reply:
Ok